File: recipes/subms-health/python/src/subms_health/env_section.py
"""Env/deploy provider: select env vars by explicit key + prefix/glob, redact
secrets, group them into a named ComponentHealth. Mirrors the Rust port exactly
(including the cross-language JSON fixture)."""

from __future__ import annotations

import os
from enum import Enum
from typing import Dict, List, Optional, Tuple

from .component import ComponentHealth, HealthIndicator
from .status import HealthStatus


class EnvProvider:
    def get(self, key: str) -> Optional[str]:  # pragma: no cover - abstract
        raise NotImplementedError

    def keys(self) -> List[str]:  # pragma: no cover - abstract
        raise NotImplementedError


class SystemEnv(EnvProvider):
    def get(self, key: str) -> Optional[str]:
        return os.environ.get(key)

    def keys(self) -> List[str]:
        return list(os.environ.keys())


class MapEnv(EnvProvider):
    def __init__(self) -> None:
        self._vars: Dict[str, str] = {}

    def with_(self, key: str, value: str) -> "MapEnv":
        self._vars[key] = value
        return self

    def get(self, key: str) -> Optional[str]:
        return self._vars.get(key)

    def keys(self) -> List[str]:
        return list(self._vars.keys())


def fnv1a(data: bytes) -> int:
    h = 0xCBF29CE484222325
    for b in data:
        h ^= b
        h = (h * 0x100000001B3) & 0xFFFFFFFFFFFFFFFF
    return h


class RedactionPolicy(Enum):
    MASK = "MASK"
    LAST4 = "LAST4"
    HASH = "HASH"
    FINGERPRINT = "FINGERPRINT"

    def apply(self, value: str) -> str:
        if self == RedactionPolicy.MASK:
            return "***"
        if self == RedactionPolicy.LAST4:
            return "***" + value[-4:] if len(value) > 4 else "***"
        if self == RedactionPolicy.HASH:
            return "fnv1a:%016x" % fnv1a(value.encode("utf-8"))
        return "fp_%06x" % (fnv1a(value.encode("utf-8")) & 0xFFFFFF)


_SECRET_NEEDLES = ["SECRET", "TOKEN", "KEY", "PASSWORD", "PASS", "CREDENTIAL"]


def _glob_match(pattern: str, s: str) -> bool:
    star = pattern.find("*")
    if star < 0:
        return pattern == s
    pre, post = pattern[:star], pattern[star + 1 :]
    return len(s) >= len(pre) + len(post) and s.startswith(pre) and s.endswith(post)


class EnvSection:
    def __init__(self, name: str) -> None:
        self._name = name
        self._explicit: List[str] = []
        self._prefixes: List[str] = []
        self._globs: List[str] = []
        self._redactions: List[Tuple[str, RedactionPolicy]] = []
        self._redact_substrings: List[Tuple[str, RedactionPolicy]] = []
        self._remap: Dict[str, str] = {}
        self._strip_prefix_in_key = False
        self._lowercase_keys = False
        self._status = HealthStatus.UP
        self._include_empty = False

    @property
    def name(self) -> str:
        return self._name

    def key(self, k: str) -> "EnvSection":
        self._explicit.append(k)
        return self

    def keys(self, ks) -> "EnvSection":
        self._explicit.extend(ks)
        return self

    def prefix(self, p: str) -> "EnvSection":
        self._prefixes.append(p)
        return self

    def glob(self, pattern: str) -> "EnvSection":
        self._globs.append(pattern)
        return self

    def redact(self, key: str, policy: RedactionPolicy) -> "EnvSection":
        self._redactions.append((key, policy))
        return self

    def redact_substring(self, needle: str, policy: RedactionPolicy) -> "EnvSection":
        self._redact_substrings.append((needle, policy))
        return self

    def redact_secrets(self) -> "EnvSection":
        for n in _SECRET_NEEDLES:
            self._redact_substrings.append((n, RedactionPolicy.MASK))
        return self

    def remap(self, frm: str, to: str) -> "EnvSection":
        self._remap[frm] = to
        return self

    def strip_prefix_in_key(self, yes: bool = True) -> "EnvSection":
        self._strip_prefix_in_key = yes
        return self

    def lowercase_keys(self, yes: bool = True) -> "EnvSection":
        self._lowercase_keys = yes
        return self

    def status(self, status: HealthStatus) -> "EnvSection":
        self._status = status
        return self

    def include_empty(self, yes: bool = True) -> "EnvSection":
        self._include_empty = yes
        return self

    def _matches(self, key: str) -> bool:
        return (
            key in self._explicit
            or any(key.startswith(p) for p in self._prefixes)
            or any(_glob_match(g, key) for g in self._globs)
        )

    def _detail_key(self, raw: str) -> str:
        if raw in self._remap:
            return self._remap[raw]
        dk = raw
        if self._strip_prefix_in_key:
            best = ""
            for p in self._prefixes:
                if raw.startswith(p) and len(p) > len(best):
                    best = p
            if best:
                dk = raw[len(best) :]
        if self._lowercase_keys:
            dk = dk.lower()
        return dk

    def _policy_for(self, raw: str, detail: str) -> Optional[RedactionPolicy]:
        for k, p in self._redactions:
            if k == raw or k == detail:
                return p
        rl, dl = raw.lower(), detail.lower()
        for s, p in self._redact_substrings:
            sl = s.lower()
            if sl in rl or sl in dl:
                return p
        return None

    def render(self, env: EnvProvider) -> ComponentHealth:
        candidates = set(self._explicit)
        for k in env.keys():
            if self._matches(k):
                candidates.add(k)
        details: Dict[str, object] = {}
        for raw in sorted(candidates):
            val = env.get(raw)
            if val is None:
                continue
            if val == "" and not self._include_empty:
                continue
            dk = self._detail_key(raw)
            policy = self._policy_for(raw, dk)
            details[dk] = policy.apply(val) if policy else val
        return ComponentHealth(self._status, details)

    def into_indicator(self, env: Optional[EnvProvider] = None) -> HealthIndicator:
        return _EnvSectionIndicator(self, env or SystemEnv())


class _EnvSectionIndicator(HealthIndicator):
    def __init__(self, section: EnvSection, env: EnvProvider) -> None:
        self._section = section
        self._env = env

    def name(self) -> str:
        return self._section.name

    def check(self) -> ComponentHealth:
        return self._section.render(self._env)

File: recipes/subms-health/java/src/main/java/com/submillisecond/recipes/health/EnvSection.java
package com.submillisecond.recipes.health;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.TreeSet;

/**
 * Env/deploy provider: select env vars by explicit key + prefix/glob, redact
 * secrets, group them into a named ComponentHealth. Mirrors the Rust + Python
 * ports (incl. the cross-language fixture).
 */
public final class EnvSection {
    private static final String[] SECRET_NEEDLES = {"SECRET", "TOKEN", "KEY", "PASSWORD", "PASS", "CREDENTIAL"};

    private final String name;
    private final List<String> explicit = new ArrayList<>();
    private final List<String> prefixes = new ArrayList<>();
    private final List<String> globs = new ArrayList<>();
    private final List<Map.Entry<String, RedactionPolicy>> redactions = new ArrayList<>();
    private final List<Map.Entry<String, RedactionPolicy>> redactSubstrings = new ArrayList<>();
    private final Map<String, String> remap = new LinkedHashMap<>();
    private boolean stripPrefixInKey = false;
    private boolean lowercaseKeys = false;
    private HealthStatus status = HealthStatus.UP;
    private boolean includeEmpty = false;

    public EnvSection(String name) {
        this.name = name;
    }

    public String name() {
        return name;
    }

    public EnvSection key(String k) {
        explicit.add(k);
        return this;
    }

    public EnvSection keys(List<String> ks) {
        explicit.addAll(ks);
        return this;
    }

    public EnvSection prefix(String p) {
        prefixes.add(p);
        return this;
    }

    public EnvSection glob(String pattern) {
        globs.add(pattern);
        return this;
    }

    public EnvSection redact(String key, RedactionPolicy policy) {
        redactions.add(Map.entry(key, policy));
        return this;
    }

    public EnvSection redactSubstring(String needle, RedactionPolicy policy) {
        redactSubstrings.add(Map.entry(needle, policy));
        return this;
    }

    public EnvSection redactSecrets() {
        for (String n : SECRET_NEEDLES) {
            redactSubstrings.add(Map.entry(n, RedactionPolicy.MASK));
        }
        return this;
    }

    public EnvSection remap(String from, String to) {
        remap.put(from, to);
        return this;
    }

    public EnvSection stripPrefixInKey(boolean yes) {
        stripPrefixInKey = yes;
        return this;
    }

    public EnvSection lowercaseKeys(boolean yes) {
        lowercaseKeys = yes;
        return this;
    }

    public EnvSection status(HealthStatus s) {
        status = s;
        return this;
    }

    public EnvSection includeEmpty(boolean yes) {
        includeEmpty = yes;
        return this;
    }

    private boolean matches(String key) {
        if (explicit.contains(key)) {
            return true;
        }
        for (String p : prefixes) {
            if (key.startsWith(p)) {
                return true;
            }
        }
        for (String g : globs) {
            if (globMatch(g, key)) {
                return true;
            }
        }
        return false;
    }

    private String detailKey(String raw) {
        if (remap.containsKey(raw)) {
            return remap.get(raw);
        }
        String dk = raw;
        if (stripPrefixInKey) {
            String best = "";
            for (String p : prefixes) {
                if (raw.startsWith(p) && p.length() > best.length()) {
                    best = p;
                }
            }
            if (!best.isEmpty()) {
                dk = raw.substring(best.length());
            }
        }
        if (lowercaseKeys) {
            dk = dk.toLowerCase(java.util.Locale.ROOT);
        }
        return dk;
    }

    private RedactionPolicy policyFor(String raw, String detail) {
        for (Map.Entry<String, RedactionPolicy> e : redactions) {
            if (e.getKey().equals(raw) || e.getKey().equals(detail)) {
                return e.getValue();
            }
        }
        String rl = raw.toLowerCase(java.util.Locale.ROOT);
        String dl = detail.toLowerCase(java.util.Locale.ROOT);
        for (Map.Entry<String, RedactionPolicy> e : redactSubstrings) {
            String sl = e.getKey().toLowerCase(java.util.Locale.ROOT);
            if (rl.contains(sl) || dl.contains(sl)) {
                return e.getValue();
            }
        }
        return null;
    }

    public ComponentHealth render(EnvProvider env) {
        TreeSet<String> candidates = new TreeSet<>(explicit);
        for (String k : env.keys()) {
            if (matches(k)) {
                candidates.add(k);
            }
        }
        ComponentHealth c = new ComponentHealth(status);
        Map<String, Object> details = c.details();
        for (String raw : candidates) {
            String val = env.get(raw);
            if (val == null) {
                continue;
            }
            if (val.isEmpty() && !includeEmpty) {
                continue;
            }
            String dk = detailKey(raw);
            RedactionPolicy policy = policyFor(raw, dk);
            details.put(dk, policy != null ? policy.apply(val) : val);
        }
        return c;
    }

    public HealthIndicator intoIndicator(EnvProvider env) {
        return new HealthIndicator() {
            @Override
            public String name() {
                return name;
            }

            @Override
            public ComponentHealth check() {
                return render(env);
            }
        };
    }

    static boolean globMatch(String pattern, String s) {
        int star = pattern.indexOf('*');
        if (star < 0) {
            return pattern.equals(s);
        }
        String pre = pattern.substring(0, star);
        String post = pattern.substring(star + 1);
        return s.length() >= pre.length() + post.length() && s.startsWith(pre) && s.endsWith(post);
    }
}
